### ISSUE TITLE
JSON variable substitution option doesn't work if json file contains Input key as top-level

### ISSUE THREAD

**Issue by rszuban** [microsoft/azure-pipelines-tasks#10589]
## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
*Type*: Bug

**Enter Task Name**: IIS Web App Deploy (Azure App Service Deploy as well, probably)

list here (V# not needed):  
[https://github.com/microsoft/azure-pipelines-tasks/tree/master/Tasks/IISWebAppDeployment](https://github.com/microsoft/azure-pipelines-tasks/tree/master/Tasks/IISWebAppDeployment)

## Environment
- Server - Azure Pipelines
- Agent - Private (single agent in a deployment group) - Windows Server 2016, Agent Version 2.150.3

## Issue Description
I'm trying to Release a package containing *appsettings.json* (among other necessary files). In the *IIS Web App Deploy* task I'm using the *JSON variable substitution* option to substitute some values in a *appsettings.json* file with release pipeline variables.

The *appsettings.json* file contains following section:

```json
{
  "Output": {
    "SomeUnimportantKey": "SomeUnimportantValue",
    "SomeOtherUnimportantKey": "SomeOtherUnimportantValue"
  },
  "Input": {
    "SomeKey": "SomeValue"
  }
}
```
To acommodate *JSON variable substitution* I created a following variable in the release pipeline *Input.SomeKey* which equals to *NewValue* (there also variables that handle modifications in *Output* section but they work as intended).

After the deployment completes values in the Output section are replaced successfully but in the Input section the *SomeKey* key is missing.

```json
{
  "Output": {
    "SomeUnimportantKey": "SomeNewUnimportantValue",
    "SomeOtherUnimportantKey": "SomeNewOtherUnimportantValue"
  },
  "Input": {}
}
```
Inspecting the agent folder structure reveals that in a *_work/rxx/a* directory there is a *temp_web_package_1731613121400375.zip* produced probably during the *JSON variable substitution* step along with *error.txt* file which has no content.

**Tl;dr:** It appears that the pipeline variable *Input.SomeKey* cannot be used for *JSON variable substitution* that contains following json `"Input": { "SomeKey": "SomeValue" }`.

As the variable name contains word **Input** it is treated as some special sort of variable and its value is unusable for *JSON variable substitution*.

### Task logs
Unfortunately due to compliance reasons I'm unable to provide full logs but upon inspecting the debug output logs I found the following.

When the IIS Web App Deploy starts there are following messages

```
2019-06-05T13:24:24.5134902Z ##[debug]agent.workFolder=C:\vstsagent\A2\_work
2019-06-05T13:24:24.5234512Z ##[debug]loading inputs and endpoints
2019-06-05T13:24:24.5244662Z ##[debug]loading ENDPOINT_AUTH_PARAMETER_SYSTEMVSSCONNECTION_ACCESSTOKEN
2019-06-05T13:24:24.5262839Z ##[debug]loading ENDPOINT_AUTH_SCHEME_SYSTEMVSSCONNECTION
2019-06-05T13:24:24.5266873Z ##[debug]loading ENDPOINT_AUTH_SYSTEMVSSCONNECTION
2019-06-05T13:24:24.5270079Z ##[debug]loading INPUT_ADDITIONALARGUMENTS
2019-06-05T13:24:24.5272516Z ##[debug]loading INPUT_EXCLUDEFILESFROMAPPDATAFLAG
2019-06-05T13:24:24.5274740Z ##[debug]loading INPUT_SOMEKEY
2019-06-05T13:24:24.5278293Z ##[debug]loading INPUT_JSONFILES
2019-06-05T13:24:24.5279693Z ##[debug]loading INPUT_PACKAGE
2019-06-05T13:24:24.5281992Z ##[debug]loading INPUT_REMOVEADDITIONALFILESFLAG
2019-06-05T13:24:24.5285647Z ##[debug]loading INPUT_SETPARAMETERSFILE
2019-06-05T13:24:24.5290039Z ##[debug]loading INPUT_TAKEAPPOFFLINEFLAG
2019-06-05T13:24:24.5291822Z ##[debug]loading INPUT_VIRTUALAPPLICATION
2019-06-05T13:24:24.5294259Z ##[debug]loading INPUT_WEBSITENAME
2019-06-05T13:24:24.5297090Z ##[debug]loading INPUT_XMLTRANSFORMATION
2019-06-05T13:24:24.5301769Z ##[debug]loading INPUT_XMLVARIABLESUBSTITUTION
2019-06-05T13:24:24.5310942Z ##[debug]loading SECRET_AGENT_PROXYPASSWORD
2019-06-05T13:24:24.5314230Z ##[debug]loading SECRET_REDACTED_FOR_COMPLIANCE
2019-06-05T13:24:24.5317077Z ##[debug]loading SECRET_REDACTED_FOR_COMPLIANCE
2019-06-05T13:24:24.5328406Z ##[debug]loaded 18
```

The pipeline variable *Input.SomeKey* is treated as a special pipeline input variable (I'm unsure by which mechanism) hence the entry *2019-06-05T13:24:24.5274740Z ##[debug]loading INPUT_SOMEKEY*.

Further in the logs there are following messages:

```
2019-06-05T13:24:29.2155783Z ##[debug]Parameters.AddBinding=True
2019-06-05T13:24:29.2158133Z ##[debug]Input.SomeKey=undefined
2019-06-05T13:24:29.2158252Z ##[debug]build.requestedForId=98a673f3-cfc4-6a52-8ccf-a8c1beea03d4
```

When the step *JSON variable substitution* is executed a following messages are shown:

```
2019-06-05T13:24:29.3364644Z ##[debug]substituting value on key: SomeUnimportantKey
2019-06-05T13:24:29.3364940Z ##[debug]substituting value on key: SomeOtherUnimportantKey
2019-06-05T13:24:29.3365145Z ##[debug]substituting value on key: SomeKey
2019-06-05T13:24:29.3469896Z JSON variable substitution applied successfully.
```

**Comment by SumiranAgg** [comment]
@rszuban Thank you for pointing out this issue. I have raised an issue with azure-pipelines-task-lib team to investigate further.

**Comment by stephenmichaelf** [comment]
Can you change it from Input to In?

**Comment by rszuban** [comment]
Unfortunately that is not possible.

**Comment by bryanmacfarlane** [comment]
Input. and Input_ are reserved as part of the protocol and that can't be changed.  Changing to a doc issue.  @stephenmichaelf - can you work with docs to ensure this is covered?

**Comment by damccorm** [comment]
Closing since this is documented here - https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch#variable-characters
